### PR TITLE
test(KFLUXVNGD-842): e2e on tekton

### PR DESCRIPTION
**Pull request by yftacherzog**
Adding Tekton tasks and pipeline for deploying the operator, optionally overriding upstream manifests, and running e2e tests.
    
This can later be used by Konflux teams to verify their changes within Konflux.

Assisted-by: Cursor

### CODE REVIEW

File: .tekton/konflux-operator-e2e-pull-request.yaml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      - name: url
+        value: '{{source_url}}'
+      - name: revision
+        value: '{{revision}}'
+      - name: pathInRepo
+        value: .tekton/pipelines/operator-e2e/pipeline.yaml
+  taskRunTemplate:
+    # kind-aws provision creates/patches Secret kfg-<PipelineRun name>; konflux-integration-runner has secrets create/patch (build-pipeline-* SAs do not).
+    serviceAccountName: konflux-integration-runner
+  workspaces:
```
[the remark this answers is not in the document]
Comment (yftacherzog) [reply .tekton/konflux-operator-e2e-pull-request.yaml]: PRs opened from forks need Tekton pipelines to be triggered with comment `/retest` to trigger this pipeline. As far as I understand it, only authorized users will be able to use this comment to trigger CI.

I think this issue extends to repos that will use the pipelines/tasks. Might be that ultimately, we need to release the pipeline/tasks as OCI artifacts.

[the remark this answers is not in the document]
Comment (yftacherzog) [reply .tekton/konflux-operator-e2e-pull-request.yaml]: BTW, I'm sure if this is very different than the issue we had with pull_request_target workflows. I.e. we accept the fact that PRs code can expose secrets, so we only allow authorized users to trigger CI.

File: .tekton/pipelines/operator-e2e/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# `operator-e2e-pipeline`
```
Comment (gbenhaim) [inline .tekton/pipelines/operator-e2e/README.md:1]: I wonder if we should include it in our docs website under the installation methods. @yftacherzog WDYT?

File: .tekton/pipelines/operator-e2e/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+- Tekton flow is `none`-mode oriented for operator install (out-of-cluster `bin/manager` in deploy task).
+- Deploy and tests are split across `deploy-konflux` and `konflux-e2e-tests`.
+
+## Inputs (params)
+
+- `git-url` (default: `https://github.com/konflux-ci/konflux-ci.git`): repository URL used for clone + git-resolved local tasks.
+- `revision` (required): git ref from `git-url` to test.
+- `overrides-yaml` (default: empty): optional inline overrides consumed by deploy task.
+- `konflux-ready-timeout` (default: `30m`): readiness timeout for Konflux CR.
+- `oci-container-repo` (default: `quay.io/example/konflux-operator-e2e-artifacts`): OCI prefix used by provision/deprovision tasks.
```
Comment (gbenhaim) [inline .tekton/pipelines/operator-e2e/README.md]: not sure I understand the purpose of this repo

File: .tekton/pipelines/operator-e2e/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- `konflux-test-infra-secret` (default: `konflux-test-infra`): secret for OCI credentials.
```
Comment (gbenhaim) [inline .tekton/pipelines/operator-e2e/README.md]: not sure I understand the purpose of this secret

File: scripts/operator-e2e/apply-overrides-from-yaml.sh [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+mkdir -p "${REPO_ROOT}/.tmp"
+if [[ -n "${OVERRIDES_YAML_PATH:-}" ]]; then
+  if [[ ! -f "${OVERRIDES_YAML_PATH}" ]]; then
+    echo "error: OVERRIDES_YAML_PATH is not a readable file: ${OVERRIDES_YAML_PATH}" >&2
+    exit 1
+  fi
+  cp "${OVERRIDES_YAML_PATH}" "${REPO_ROOT}/.tmp/overrides.yaml"
+else
+  : "${OVERRIDES_YAML:?set OVERRIDES_YAML or OVERRIDES_YAML_PATH}"
+  echo "${OVERRIDES_YAML}" > "${REPO_ROOT}/.tmp/overrides.yaml"
```
Comment (gbenhaim) [inline scripts/operator-e2e/apply-overrides-from-yaml.sh]: 1. echo can misinterpret YAML content — apply-overrides-from-yaml.sh line 33
Severity: Concern

echo "${OVERRIDES_YAML}" will interpret leading -n, -e, or -E as options on some implementations (e.g. dash, some busybox). Since OVERRIDES_YAML is user-supplied YAML that starts with - name:, the leading - could trigger this.


apply-overrides-from-yaml.sh
Lines 33-33
  echo "${OVERRIDES_YAML}" > "${REPO_ROOT}/.tmp/overrides.yaml"
Suggested fix:


printf '%s\n' "${OVERRIDES_YAML}" > "${REPO_ROOT}/.tmp/overrides.yaml"

File: .tekton/pipelines/operator-e2e/pipeline.yaml [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            '[{\"containerPort\":30012,\"hostPort\":8180,\"protocol\":\"TCP\"}]'
+        - name: cpus
+          value: "16"
+        - name: memory
+          value: "32"
+        - name: spot
+          value: "false"
+        - name: harden-control-plane
+          value: "true"
+    - name: clone-repository
```
Comment (gbenhaim) [inline .tekton/pipelines/operator-e2e/pipeline.yaml]: any reason not to use the git-clone task konflux provides?

File: .tekton/konflux-operator-e2e-pull-request.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+metadata:
+  annotations:
+    build.appstudio.openshift.io/repo: https://github.com/konflux-ci/konflux-ci?rev={{revision}}
+    build.appstudio.redhat.com/commit_sha: '{{revision}}'
+    build.appstudio.redhat.com/pull_request_number: '{{pull_request_number}}'
+    build.appstudio.redhat.com/target_branch: '{{target_branch}}'
+    pipelinesascode.tekton.dev/cancel-in-progress: "true"
+    pipelinesascode.tekton.dev/max-keep-runs: "2"
+    pipelinesascode.tekton.dev/on-cel-expression: |
+      event == "pull_request" && target_branch == "main" ||
```
Comment (avi-biton) [inline .tekton/konflux-operator-e2e-pull-request.yaml:12]: Do we want to run it for each PR in main?
We can reduce it only to operator related PRs and also  for example filter out documentation PRs

File: .tekton/konflux-operator-e2e-pull-request.yaml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      # value: |
+      #   - name: segment-bridge
+      #     git:
+      #       - sourceRepo: konflux-ci/segment-bridge
+      #         remote:
+      #           repo: https://github.com/konflux-ci/segment-bridge
+      #           ref: {{revision}}
+      #     images:
+      #       - orig: quay.io/konflux-ci/segment-bridge
+      #         replacement: quay.io/redhat-user-workloads/konflux-vanguard-tenant/segment-bridge/segment-bridge:on-pr-{{revision}}
```
Comment (avi-biton) [inline .tekton/konflux-operator-e2e-pull-request.yaml]: The `revision` parameter is taken from `konflux-ci` sha.
It does not necessary means that we have the same sha in other repositories (like `segment-bridge` in this example)

Comment (yftacherzog) [reply .tekton/konflux-operator-e2e-pull-request.yaml]: It's coming from PAC (https://pipelinesascode.com/docs/concepts/#8-variable-substitution)

Comment (avi-biton) [reply .tekton/konflux-operator-e2e-pull-request.yaml]: So IIUC, this example is valid for use in components repository where the revision will be taken from their sha and will be replaced correctly, however, if we will try to use overrides in this repository it will fail

Comment (yftacherzog) [reply .tekton/konflux-operator-e2e-pull-request.yaml]: We can use it everywhere, but the content of the overrides will be different. When running in our repo, I don't think there's a practical use case for the `revision` argument, but for our upstream services, I think it will be the main instrument for testing PR code together with latest from konflux-ci/konflux-ci.

File: .tekton/konflux-operator-e2e-pull-request.yaml (hunk 7 above)
Comment (gbenhaim) [reply .tekton/konflux-operator-e2e-pull-request.yaml:12]: I thinl that it can be beneficial to start with running it for each pr so we can understand how stable it's.

File: .tekton/pipelines/operator-e2e/pipeline.yaml (hunk 6 above)
Comment (yftacherzog) [reply .tekton/pipelines/operator-e2e/pipeline.yaml]: I'll look into that.

File: scripts/operator-e2e/apply-overrides-from-yaml.sh (hunk 5 above)
Comment (yftacherzog) [reply scripts/operator-e2e/apply-overrides-from-yaml.sh]: I'm testing a go-based approach. I'll close this one for now.

File: .tekton/pipelines/operator-e2e/README.md (hunk 2 above)
Comment (yftacherzog) [reply .tekton/pipelines/operator-e2e/README.md:1]: Might be

File: .tekton/pipelines/operator-e2e/pipeline.yaml (hunk 6 above)
Comment (yftacherzog) [reply .tekton/pipelines/operator-e2e/pipeline.yaml]: Changed

File: .tekton/pipelines/operator-e2e/README.md (hunk 3 above)
Comment (yftacherzog) [reply .tekton/pipelines/operator-e2e/README.md]: It's used by the VM provisioning process to store logs. I'm removing the default value as it makes no sense to have it.

File: .tekton/pipelines/operator-e2e/README.md (hunk 4 above)
Comment (yftacherzog) [reply .tekton/pipelines/operator-e2e/README.md]: I renamed some of the things, changed the default values and enhanced the pipeline readme to explain how the secret should look like.

**Comment by yftacherzog** [comment]
@gbenhaim @avi-biton @flacatus 
The pipeline is now pushing the manifests that are used for deploying Konflux to Quay, so we have more confidence in what is being deployed when manifests are overridden by the pipeline. For example, below, is the result of a run in which I overridden segment and ui, and also included a broken ref to an image to see that Konflux is failing deployment.

This is the artifact: quay.io/konflux-vanguard/konflux-operator-e2e-artifacts:konflux-operator-e2e-on-pull-request-62qnh.pkg-manifests

This is the overrides inputs:

```yaml
    - name: overrides-yaml
      value: |
        - name: segment-bridge
          git:
            - sourceRepo: konflux-ci/segment-bridge
              remote:
                repo: https://github.com/konflux-ci/segment-bridge
                ref: f9b0012a9b90842c502dc88419000a6eabb9b276
          images:
            - orig: quay.io/konflux-ci/segment-bridge
              replacement: quay.io/redhat-user-workloads/konflux-vanguard-tenant/segment-bridge/segment-bridge:on-pr-f9b0012a9b90842c502dc88419000a6eabb9b276
        - name: ui
          images:
            - orig: quay.io/konflux-ci/konflux-ui
              replacement: quay.io/redhat-user-workloads/konflux-ui-tenant/konflux-ui/konflux-ui:on-pr-e8b4f4148e19ba5a456bc8786d9b7d81a48fd1ccZZZ
```

File: operator/pkg/overrides/overrides.go [hunk 9]
```diff
@@ -0,0 +1,660 @@
...
+				if !ok {
+					continue
+				}
+				if imageNameFromMap(im) != ov.Orig {
+					continue
+				}
+				im["newName"] = newName
+				im["newTag"] = newTag
+				delete(im, "digest")
+				changed = true
```
[the remark this answers is not in the document]
Comment (yftacherzog) [reply operator/pkg/overrides/overrides.go:424]: This should now be fixed. can you reevaluate?

[the remark this answers is not in the document]
Comment (yftacherzog) [reply operator/pkg/overrides/overrides.go:424]: /ask "This should now be fixed. can you reevaluate?"

File: operator/cmd/overrides/main.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
+package main
```
Comment (gbenhaim) [inline operator/cmd/overrides/main.go:11]: is it possible to add module docs and describe the purpose of this command? it would be useful for AI as well.

**Comment by gbenhaim** [comment]
_Comment from Gal's Cursor_

It would be useful to document somewhere (pipeline README or `scripts/operator-e2e/README.md`) that the operator process (`bin/manager`) **terminates when the deploy-konflux Task completes** — it runs as a background process inside that Task's pod, so when the pod exits, the operator is gone.

The `konflux-e2e-tests` Task then runs in a separate pod **without the operator running**. The tests rely on the services the operator deployed (build-service, integration-service, PaC, etc.) which continue running independently in the Kind cluster, but there is no operator reconciliation loop active during the test phase.

This is fine for the current conformance suite since it exercises the deployed services rather than operator reconciliation. But it's a non-obvious aspect of the architecture that future contributors should be aware of — e.g., if a test were to delete a Deployment expecting the operator to recreate it, that wouldn't work.

A short note in the pipeline README's "Scope" section (which already mentions `none` mode) would be a natural place for this.

**Comment by gbenhaim** [comment]
_Comment from Gal's Cursor_

### Suggestion: additional test coverage for `operator/pkg/overrides`

The existing 14 tests cover the happy paths well, but a few areas could use more coverage:

**1. Validation edge cases (`validateOverrides`)**
Currently only one rejection case is tested. Worth adding negative tests for:
- Both `Remote` and `LocalPath` set on the same `GitRule`
- Missing `SourceRepo`
- Invalid `SourceRepo` format (not `org/repo`)
- Invalid image references
- Duplicate component names

**2. Input parsing helpers — no direct tests today**
- `parseImageReference` — handles `name:tag@digest` parsing; only `splitImageReference` is tested
- `normalizeOrgRepo` / `githubURLParts` — parse GitHub URLs and org/repo strings, edge cases include trailing slashes, `.git` suffixes, HTTPS vs SSH
- `firstMatchingRule` — matches git rules to kustomization resource URLs

These are all parsing external/user input, so they're high-value targets for table-driven tests.

**3. `Runner.Apply()` with git rules**
`TestRunnerApplyWithImageOverrideOnly` only exercises the image override path. There's no integration-level test that goes through git rule application → `kustomize build` → manifest rebuild. (Understandable if `kustomize` availability is a concern — could be gated with a build tag or test skip.)

Not a blocker, but would increase confidence in the override engine — especially since the `overrides-yaml` parameter is user-facing input from PipelineRun params.

File: scripts/operator-e2e/prepare-conformance-env.sh [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+append_env() {
+  local line="$1"
+  if [[ -n "${GITHUB_ENV:-}" ]]; then
+    echo "$line" >> "$GITHUB_ENV"
+  else
+    echo "export $line"
+  fi
+}
+
+line="$(grep 'RELEASE_SERVICE_CATALOG_REVISION=' test/e2e/release-service-catalog-revision | sed 's/^export //')"
```
Comment (gbenhaim) [inline scripts/operator-e2e/prepare-conformance-env.sh]: we don't have this file anymore. the revision is baked into the setup-release.sh script.

File: .tekton/tasks/konflux-e2e-tests/task.yaml [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+          valueFrom:
+            secretKeyRef:
+              name: konflux-operator-e2e-credentials
+              key: GH_ORG
+        - name: GH_TOKEN
+          valueFrom:
+            secretKeyRef:
+              name: konflux-operator-e2e-credentials
+              key: GH_TOKEN
+        - name: QUAY_DOCKERCONFIGJSON
```
Comment (gbenhaim) [inline .tekton/tasks/konflux-e2e-tests/task.yaml]: QUAY_DOCKERCONFIGJSON and RELEASE_CATALOG_TA_QUAY_TOKEN are not needed anymore, since we use image-controller for creating the repositories required for the release process. it's done as part of setup-release.sh.

**Comment by gbenhaim** [comment]
## Conformance test setup is out of sync with `main`

The PR's conformance test code diverges from the current `main` branch in several significant ways. It appears to be based on an older version of the tests that predates the `setup-release.sh` consolidation.

### 1. Release setup mechanism: `setup-release.sh` ConfigMap (main) vs. inline Go code (PR)

**On `main`**, the conformance test calls `runSetupRelease()` which downloads `setup-release.sh` from the `konflux-cli/setup-release` ConfigMap (deployed by the operator) and executes it. This single script handles creating the managed namespace, ImageRepositories, EnterpriseContractPolicy, ReleasePlanAdmission, etc.

**In this PR**, the conformance test calls `createReleaseConfig()` — a completely different Go function that manually creates all release resources via Kubernetes client calls. This bypasses the operator-shipped script and re-implements its logic in Go.

### 2. Catalog revision source: deleted file (PR) vs. `setup-release.sh` default (main)

**On `main`**: The catalog revision is baked into `setup-release.sh` (line 87: `CATALOG_REVISION="48a31f6..."`) and tracked by Renovate there. No external file is needed.

**In this PR**: `prepare-conformance-env.sh` (line 18) reads `RELEASE_SERVICE_CATALOG_REVISION` from `test/e2e/release-service-catalog-revision`:
```bash
line="$(grep 'RELEASE_SERVICE_CATALOG_REVISION=' test/e2e/release-service-catalog-revision | sed 's/^export //')"
```
And `GetReleaseServiceCatalogRevision()` in Go also falls back to that same file.

**That file no longer exists on `main`.** Both paths will fail at runtime.

### 3. Release pipeline path: different pipelines

- **`main`**: `setup-release.sh` uses `pipelines/managed/push-to-external-registry/push-to-external-registry.yaml`
- **PR**: `createReleaseConfig()` uses `pipelines/managed/e2e/e2e.yaml`

### Suggested fix

The PR should rebase on current `main` and use the existing `runSetupRelease()` + `patchECPForE2E()` pattern from `main`'s `setup.go`, which delegates to the operator-shipped `setup-release.sh` ConfigMap. This would:
- Eliminate the need for `prepare-conformance-env.sh` to resolve the catalog revision (it's already in the script)
- Use the same release pipeline as the rest of CI
- Avoid re-implementing release resource creation in Go

---
*From Gal's Cursor*

**Comment by gbenhaim** [comment]
## Question: custom test suite support

The existing `test/e2e/run-e2e.sh` forwards extra arguments (`"$@"`) to `go test`, allowing callers to pass `-ginkgo.focus`, `-ginkgo.skip`, etc. The new Tekton pipeline hardcodes the test invocation without exposing any such parameter.

Could you double-check whether the legacy upstream task being replaced here allowed users to provide their own test suite or test filters? If so, it would be good to preserve that capability (e.g., via an optional pipeline param).

---
*From Gal's Cursor*

**Comment by gbenhaim** [comment]
## Suggestion: make the release service catalog revision customizable

Currently the pipeline hardcodes / auto-extracts the `RELEASE_SERVICE_CATALOG_REVISION` from the deployed manifests. It would be valuable to expose this as an optional pipeline parameter (with the current behavior as the default).

This would allow the `konflux-ci/release-service-catalog` repo to reuse this same Tekton pipeline in its own CI — pointing the revision at the PR branch under test instead of the pinned production SHA. Without this, that repo would need a separate test harness or fork the pipeline definition.

---
*From Gal's Cursor*

**Comment by yftacherzog** [comment]
> Could you double-check whether the legacy upstream task being replaced here allowed users to provide their own test suite or test filters? If so, it would be good to preserve that capability (e.g., via an optional pipeline param).
> 
> _From Gal's Cursor_

The legacy task only deployed Konflux, so for the task running the e2e tests, we can decide what to include.

File: operator/cmd/overrides/main.go (hunk 10 above)
Comment (yftacherzog) [reply operator/cmd/overrides/main.go:11]: Done

File: scripts/operator-e2e/prepare-conformance-env.sh (hunk 11 above)
Comment (yftacherzog) [reply scripts/operator-e2e/prepare-conformance-env.sh]: I rebased on top of the latest changes and aligned with the current approach.

File: .tekton/tasks/konflux-e2e-tests/task.yaml (hunk 12 above)
Comment (yftacherzog) [reply .tekton/tasks/konflux-e2e-tests/task.yaml]: done

**Comment by yftacherzog** [comment]
> ## Suggestion: make the release service catalog revision customizable
> Currently the pipeline hardcodes / auto-extracts the `RELEASE_SERVICE_CATALOG_REVISION` from the deployed manifests. It would be valuable to expose this as an optional pipeline parameter (with the current behavior as the default).
> 
> This would allow the `konflux-ci/release-service-catalog` repo to reuse this same Tekton pipeline in its own CI — pointing the revision at the PR branch under test instead of the pinned production SHA. Without this, that repo would need a separate test harness or fork the pipeline definition.
> 
> _From Gal's Cursor_

I'm tempted to leave this for later.

**Comment by yftacherzog** [comment]
> #### High
> * **[Correctness]** `.tekton/pipelines/operator-e2e/pipeline.yaml:416,480` — ITS (`deploy-konflux-its` and `konflux-e2e-tests-its`) task resolver `revision` is hardcoded to `upstream-ci-action` (the PR branch) instead of `main`. Once the branch is deleted after merge, the ITS flow will fail to resolve task YAML.
>   Remediation: Change both `value: upstream-ci-action` lines to `value: main`.

this is temporary so we can test the PR changes.

**Comment by yftacherzog** [comment]
>[Possible Issue](https://github.com/konflux-ci/konflux-ci/pull/6157/files#diff-6d96153337934071a24af2fda56a1265eeb8579270f97bec1c23502375336c3aR257-R337)
applyGitRulesToKustomization updates images[].newTag for all images whenever exactly one remote ref is used in rewritten resources. This can unintentionally retag unrelated images in the same kustomization file. Consider scoping the newTag update to images that correspond to the rewritten git resource/repo, or avoid mutating images automatically unless an explicit image override exists.

Added a test case and applied a fix.

**Comment by yftacherzog** [comment]
@gbenhaim I addressed your comments except for one which I prefer we keep for later. I also went over the bots reviews and addressed the points which I think are important at the moment.

The pipeline is referencing the PR branch as the source of the tasks to allow it to be tested. This will need to change to point to `main` before we merge this PR. Users could pin the task reference by digest.

**Comment by gbenhaim** [comment]
Lgtm

**Comment by gbenhaim** [comment]
What is this konfku-ci-review bot? Can I configured to reuse the same comment when updating the pr?

**Comment by yftacherzog** [comment]
> What is this konfku-ci-review bot? Can I configured to reuse the same comment when updating the pr?

That's the review from the Fullsend process. We have it more complete on the caching repo.
@ralphbean @kelchen123 @amisstea can comment on what's possible here in terms of configs.

**Comment by yftacherzog** [comment]
> Lgtm

@gbenhaim I changed the pipeline to point to main now. It's expected to fail now until we have the tekton resources on that branch.

**Review by gbenhaim — APPROVED** [review]

**Review by flacatus — APPROVED** [review]
